examples/de-emphasize.rs

```diff
@@ -0,0 +1,88 @@
+extern crate mdbook;
+extern crate pulldown_cmark;
+extern crate pulldown_cmark_to_cmark;
+
+// This program removes all forms of emphasis from the markdown of the book.
+
+use mdbook::errors::Error;
+use mdbook::MDBook;
+use mdbook::book::{Book, BookItem};
+use mdbook::preprocess::{Preprocessor, PreprocessorContext};
+use pulldown_cmark::{Event, Parser, Tag};
+use pulldown_cmark_to_cmark::fmt::cmark;
+
+use std::ffi::OsString;
+use std::env::{args, args_os};
+use std::process;
+
+struct Deemphasize;
+
+impl Preprocessor for Deemphasize {
+    fn name(&self) -> &str {
+        "md-links-to-html-links"
+    }
+
+    fn run(&self, _ctx: &PreprocessorContext, book: &mut Book) -> ::std::result::Result<(), Error> {
+        eprintln!("Running '{}' preprocessor", self.name());
+        let mut res: Option<_> = None;
+        let mut num_removed_items = 0;
+        book.for_each_mut(|item: &mut BookItem| {
+            if let &Some(Err(_)) = &res {
```

Comment:
Just seeing this makes me think it'd be nicer to make the [`Book::sections`][1] field public and let people recursively walk the book's contents. The reason I originally wanted to have a `for_each_mut()` method is to prevent accidental iterator invalidation-like issues (e.g. if I add a nested item to some chapter, should we visit that too?) or forgetting to recurse.

Looking at how painful it is to continue on or return early, this approach may be a bit too restrictive... Thoughts?

[1]: http://rust-lang-nursery.github.io/mdBook/src/mdbook/book/book.rs.html#74

Reply:
Indeed! These were my thoughts too. I was wondering too as to why these fields are hidden - they making interacting with the book more difficult on both sides. But on the other hand I am someone who is a sworn enemy of data hiding in case the internal state is fully understood.

My suggestion would be to just make the state of the mdbook public, as there is no 'invalid' state thanks to Rusts type system. There are no special variants that would have to be maintained by methods, for example.

As a sidenote: I noticed that it's impossible to currently get rid of the default renderer, which is an inconvenience for `termbook play` for example, as it will always generate html docs as side-effect.